pkg/webhook/user/mutation/mutation.go

```diff
@@ -0,0 +1,176 @@
+			return webhook.Errored(http.StatusBadRequest, err)
+		}
+
+	case admissionv1.Delete:
+		return webhook.Allowed(fmt.Sprintf("no mutation done for request %s", req.UID))
+
+	default:
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("%s not supported on user resources", req.Operation))
+	}
+
+	mutatedKey, err := json.Marshal(newUser)
+	if err != nil {
+		return webhook.Errored(http.StatusInternalServerError, fmt.Errorf("marshaling user object failed: %w", err))
+	}
+
+	return admission.PatchResponseFromRaw(req.Object.Raw, mutatedKey)
+}
+
+func (h *AdmissionHandler) ensureProjectRelation(ctx context.Context, user *kubermaticv1.User, oldUser *kubermaticv1.User) error {
+	isUpdate := oldUser != nil
+
+	if isUpdate && user.Spec.Project != oldUser.Spec.Project {
+		return errors.New("cannot change the project for an service account User object")
+	}
+
+	if kubermaticv1helper.IsProjectServiceAccount(user.Spec.Email) {
+		if user.Spec.Project == "" {
+			return errors.New("project name must be configured")
+		}
+
+		project := &kubermaticv1.Project{}
+		if err := h.client.Get(ctx, types.NamespacedName{Name: user.Spec.Project}, project); err != nil {
+			if kerrors.IsNotFound(err) {
+				// during user creation, we enforce the project association;
+				// during updates we are more relaxed and only require that the association isn't changed,
+				// so that if a project gets removed before the user (for whatever reason), then
+				// the cluster cleanup can still progress and is not blocked by the webhook
+				if isUpdate {
+					return nil
+				}
+
+				return errors.New("no such project exists")
+			}
+
+			return fmt.Errorf("failed to get project: %w", err)
+		}
+
+		// Do not check the project phase, as projects only get Active after being successfully
+		// reconciled. This requires the owner user to be setup properly as well, which in turn
+		// requires owner references to be setup. All of this is super annoying when doing
+		// GitOps. Instead we rely on _eventual_ consistency and only check that the project
+		// exists and is not being deleted.
+		if !isUpdate && project.DeletionTimestamp != nil {
+			return errors.New("project is in deletion, cannot create new service accounts in it")
+		}
+
+		// ensure the user has exactly 1 OwnerRef to a Project and that this ownerRef points to the
+		// correct Project
+		ownerRefs := []metav1.OwnerReference{}
+		apiVersion := kubermaticv1.SchemeGroupVersion.String()
+		kind := kubermaticv1.ProjectKindName
+
+		for _, ref := range user.OwnerReferences {
+			// skip all project owner refs
+			if ref.Kind == kind && ref.APIVersion == apiVersion {
+				continue
+			}
+
+			ownerRefs = append(ownerRefs, ref)
+		}
+
+		ownerRefs = append(ownerRefs, metav1.OwnerReference{
+			APIVersion: apiVersion,
+			Kind:       kind,
+			UID:        project.UID,
+			Name:       project.Name,
+		})
+
+		kuberneteshelper.SortOwnerReferences(ownerRefs)
+		user.SetOwnerReferences(ownerRefs)
```

Comment:
We did this in one or 2 more places, and I expect we could use it again somewhere. Maybe it would be useful to extract this to a function?